packages/common/core-interfaces/src/exposedUtilityTypes.ts

```diff
@@ -0,0 +1,565 @@
+					? /* primitive or replaced types => */ T
+					: /* test for known types that become null */ T extends
+								| undefined
+								| symbol
+								// eslint-disable-next-line @typescript-eslint/ban-types
+								| Function
+						? /* => */ null
+						: TBlessed;
+
+	/**
+	 * Checks for a type that is simple class of number and string indexed types to numbers and strings.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type IsEnumLike<T extends object> = T extends readonly (infer _)[]
+		? /* array => */ false
+		: T extends {
+					// all numerical indices should refer to a string
+					readonly [i: number]: string;
+					// string indices may be string or number
+					readonly [p: string]: number | string;
+				}
+			? /* test for a never or any property */ true extends {
+					[K in keyof T]: T[K] extends never ? true : never;
+				}[keyof T]
+				? false
+				: true
+			: false;
+
+	/**
+	 * Test for type equality
+	 *
+	 * @returns IfSame if identical and IfDifferent otherwise.
+	 *
+	 * Implementation derived from https://github.com/Microsoft/TypeScript/issues/27024#issuecomment-421529650
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type IfSameType<X, Y, IfSame = unknown, IfDifferent = never> = (<T>() => T extends X
+		? 1
+		: 2) extends <T>() => T extends Y ? 1 : 2
+		? IfSame
+		: IfDifferent;
+
+	/**
+	 * Test for type equality
+	 *
+	 * @returns `true` if identical and `false` otherwise.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type IsSameType<X, Y> = IfSameType<X, Y, true, false>;
+
+	/**
+	 * Checks that type is exactly `object`.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type IsExactlyObject<T extends object> = IsSameType<T, object>;
+
+	/**
+	 * Creates a simple object type from an intersection of multiple.
+	 * @privateRemarks `T extends Record` encourages tsc to process intersections within unions.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type FlattenIntersection<T> = T extends Record<string | number | symbol, unknown>
+		? {
+				[K in keyof T]: T[K];
+			}
+		: T;
+
+	/**
+	 * Replaces any instance where a type T recurses into itself or a portion of
+	 * itself with TReplacement.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type ReplaceRecursionWith<T, TReplacement> = ReplaceRecursionWithImpl<
+		T,
+		TReplacement,
+		never
+	>;
+
+	/**
+	 * Implementation for {@link InternalUtilityTypes.ReplaceRecursionWith}
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type ReplaceRecursionWithImpl<T, TReplacement, TAncestorTypes> =
+		/* test for recursion */ T extends TAncestorTypes
+			? /* recursion => use replacement */ TReplacement
+			: T extends object
+				? // eslint-disable-next-line @typescript-eslint/ban-types
+					/* test for function */ T extends Function
+					? /* function => */ T
+					: /* property bag => */ {
+							[K in keyof T]: ReplaceRecursionWithImpl<T[K], TReplacement, TAncestorTypes | T>;
+						}
+				: /* non-object => T as is */ T;
+
+	/**
+	 * Test for non-public properties (class instance type)
```

Comment:
I played with this a bit. Its surprisingly robust. Neat utility.

```suggestion
	 * Test for non-public properties (which can only exist on class instance types).
	 * Returns true if `T` deeply may contain (via non function members) a private or protected field.
	 * @remarks
	 * This does not distribute over unions: `ReplaceRecursionWith<A> | ReplaceRecursionWith<B>` will produce Boolean when `ReplaceRecursionWith<A | B>` produces true;
```

I did however find at least one edge case it doesn't handle: classes which extend Function return false even if they have private members. I didn't try out recursive cases.

Might be worth calling out that this has no impact on properties which are present on objects, but not included in the type. That may be obvious, but with the ways this is used, that limitation is likely to impact people. If I implement an interface with a class, or up cast a more specific type to a more general interface before it goes into the JsonSeralizable logic, it all fails to detect the omitted members despite them still impacting runtime. This limitation seems more likely to cause issues most of the things it does detect, which makes me wonder if this whole mechanism is really worth all the complexity as there is no way to fix that issue.

Reply:
I've added notes. Hopefully it will help people working quick and not being careful about classes. Use of interfaces and hiding classes takes extra effort and my experience says that is not where people start. I became particularly concerned with this issue looking at Bohemia codebase and took a stab at it.

I think I have covered the additional cases and made various improvements as I went.